mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td

```diff
@@ -411,10 +411,188 @@ def SYCLMemref : AnyTypeOf<[
 def IndexType : AnyTypeOf<[I32, I64, Index]>;
 def SYCLGetIDResult : AnyTypeOf<[I64, SYCL_IDType]>;
 def SYCLGetRangeResult : AnyTypeOf<[I64, SYCL_RangeType]>;
+def SYCLIndexSpaceGetIDResult : AnyTypeOf<[Index, SYCL_IDType]>;
+def SYCLIndexSpaceGetRangeResult : AnyTypeOf<[Index, SYCL_RangeType]>;
 def SYCLGetResult : AnyTypeOf<[I64, MemRefOf<[I64]>]>;
 def VectorEltTy : AnyTypeOf<[I1, I8, I16, I32, I64, F16, F32, F64]>;
 def VectorSplatArg : MemRefOf<[VectorEltTy]>;
 
+////////////////////////////////////////////////////////////////////////////////
+// SYCL GRID OPERATIONS
+////////////////////////////////////////////////////////////////////////////////
+
+def SYCLNumWorkItems : SYCL_Op<"num_work_items", [SYCLIndexSpaceGetRange]> {
+  let summary = "Retrieve the number of work-items.";
+  let description = [{
+    This operation returns the number of work-items in the index space. If the
+    optional argument is passed, the number of work-items in the given dimension
+    is returned.
+  }];
+
+  let arguments = (ins Optional<I32>:$dimension);
```

Comment:
I know this matches the SYCL specification, but I'm wondering if using an `i32` is the best choice here. 

There is really only a small number of allowed values (currently, 0, 1, and 2), and we probably want to detect cases where a different value is used before lowering to these operations.
So we could have an `EnumAttr` instead, to make reasoning about the dimension in the lowering/matching easier.

Reply:
I used `I32` as it enabled the use of arbitrary values here (as the spec says) and also it would be friendlier for codegen (we'd need to have constant values otherwise). Lowering to GPU dialect might be performed as follows:
1. If the input is constant lower to the provided `x`, `y`, `z` GPU attributes
2. Otherwise, we'd have to lower to code checking the value at runtime.

Reply:
What code does DPC++ currently generate if the dimension is not constant?

Reply:
It builds a full `nd_range<dim>` and uses `getelementptr` assuming the argument is in the range `[0, 3)`.

Reply:
After some considerations, I think `i32` should be fine. I agree with the point that it makes codegen easier, in case the dimension is not coming directly from some constant and you would need to follow def-use.